update: support Qwen2-57B-A14B

legraphista opened the pull request:
Added support for keys `moe_intermediate_size` and `shared_expert_intermediate_size` for Qwen2-57B-A14B

caveat:
since `self.gguf_writer.add_feed_forward_length` was getting called by `super().set_gguf_parameters()`, I had to copy-paste the existing code to update it properly

legraphista committed:
update: convert-hf-to-gguf.py to support Qwen2-57B-A14B

CISC commented:
You need to fix up `llama.cpp` too otherwise sizes will be wrong again.

Specifically, here:
https://github.com/ggerganov/llama.cpp/blob/c9ee7118d5644dd3df70ea6878b36a9761616aab/llama.cpp#L5808
do the same as `DeepSeekV2` here:
https://github.com/ggerganov/llama.cpp/blob/c9ee7118d5644dd3df70ea6878b36a9761616aab/llama.cpp#L4484
and here:
https://github.com/ggerganov/llama.cpp/blob/c9ee7118d5644dd3df70ea6878b36a9761616aab/llama.cpp#L5017

legraphista commented:
> You should fix up `llama.cpp` too.
> 
> Specifically, here:
> 
> https://github.com/ggerganov/llama.cpp/blob/c9ee7118d5644dd3df70ea6878b36a9761616aab/llama.cpp#L5808
> 
> 
> do the same as `DeepSeekV2` here:
> https://github.com/ggerganov/llama.cpp/blob/c9ee7118d5644dd3df70ea6878b36a9761616aab/llama.cpp#L4484
> 
> 
> and here:
> https://github.com/ggerganov/llama.cpp/blob/c9ee7118d5644dd3df70ea6878b36a9761616aab/llama.cpp#L6352
> 
> 
> and here:
> https://github.com/ggerganov/llama.cpp/blob/c9ee7118d5644dd3df70ea6878b36a9761616aab/llama.cpp#L5017

I'm sorry, as I'm not that well versed in this topic, can you please explain why those changes need to happen? I'm already setting the feed-forward size from the `moe_intermediate_size` param.

CISC commented:
> I'm sorry, as I'm not that well versed in this topic, can you please explain why those changes need to happen? I'm already setting the expert feed-forward size from the `moe_intermediate_size` param and the feed-forward size from `shared_expert_intermediate_size`.

In the first line I referenced you can see it doesn't actually read the expert feed-forward size you are setting, it's calculated from feed-forward and `n_experts_used`. Right now that's not an issue as the result will be the same, however that leaves setting the expert feed-forward moot.

For backwards and forwards compatibility you need to read that value from metadata if it exists, and do the calculation if it doesn't.

legraphista commented:
> In the first line I referenced you can see it doesn't actually read the expert feed-forward size you are setting, it's calculated from feed-forward and `n_experts_used`. Right now that's not an issue as the result will be the same, however that leaves setting the expert feed-forward moot.
> 
> For backwards and forwards compatibility you need to read that value from metadata if it exists, and do the calculation if it doesn't.

Ah, I see. Thank you for the explanation. I will add the changes as soon as I get an opportunity.

legraphista committed:
fix: QWEN2MOE support for expert_feed_forward_length

previously, expert ff was taken from n_ff (intermediate size) but it is now properly taken from LLM_KV_EXPERT_FEED_FORWARD_LENGTH

n_ff_exp and n_ff_shared_exp are now properly calculated

legraphista commented:
I've pushed the llama.cpp changes as requested & cleanup up the convert-hf-to-gguf.py file, as the workaround was not needed anymore

legraphista commented on llama.cpp:
```diff
@@ -5019,6 +5021,10 @@ static void llm_load_print_meta(llama_model_loader & ml, llama_model & model) {
     }
+
+    if (model.arch == LLM_ARCH_QWEN2MOE) {
+        LLAMA_LOG_INFO("%s: n_ff_exp         = %d\n",     __func__, hparams.n_ff_exp);
```
it is currently aligned with the rest of the log:
```
llm_load_print_meta: general.name     = Qwen2-57B-A14B-Instruct
llm_load_print_meta: BOS token        = 151643 '<|endoftext|>'
llm_load_print_meta: EOS token        = 151645 '<|im_end|>'
llm_load_print_meta: PAD token        = 151643 '<|endoftext|>'
llm_load_print_meta: LF token         = 148848 'ÄĬ'
llm_load_print_meta: EOT token        = 151645 '<|im_end|>'
llm_load_print_meta: n_ff_exp         = 2560                            # <- this line
```

it didn't make sense for me to align it to the `LLM_ARCH_DEEPSEEK2` branch as that code is not called, and it would look unaligned:
```
llm_load_print_meta: general.name     = Qwen2-57B-A14B-Instruct
llm_load_print_meta: BOS token        = 151643 '<|endoftext|>'
llm_load_print_meta: EOS token        = 151645 '<|im_end|>'
llm_load_print_meta: PAD token        = 151643 '<|endoftext|>'
llm_load_print_meta: LF token         = 148848 'ÄĬ'
llm_load_print_meta: EOT token        = 151645 '<|im_end|>'
llm_load_print_meta: n_ff_exp             = 2560                            # <- this line
```

ggerganov commented:
Fix the lint checks (EditorConfig and flake8)

@CISC could you review/test the changes?

legraphista committed:
update: convert-hf-to-gguf.py cleanup for Qwen2MoeForCausalLM

CISC commented on convert-hf-to-gguf.py:
```diff
@@ -1624,54 +1624,13 @@ class Qwen2MoeModel(Model):
 
-        if (shared_expert_intermediate_size := self.find_hparam(["shared_expert_intermediate_size","intermediate_size", "n_inner"])) is not None:
-            self.gguf_writer.add_feed_forward_length(shared_expert_intermediate_size)
-            logger.info(f"gguf: feed forward length = {shared_expert_intermediate_size}")
```
Why was this removed?

legraphista replied:
not removed, but cleaned up. 
`add_feed_forward_length` is called by `super().set_gguf_parameters()` as it was before my PR. 

`add_expert_feed_forward_length` is the only new thing that is added, and is the one taken into account here https://github.com/ggerganov/llama.cpp/pull/7835/files#diff-150dc86746a90bad4fc2c3334aeb9b5887b3adad3cc1459446717638605348efR5814 and here https://github.com/ggerganov/llama.cpp/pull/7835/files#diff-150dc86746a90bad4fc2c3334aeb9b5887b3adad3cc1459446717638605348efR5820

CISC replied:
Not sure as what though, @ggerganov any thoughts?

Trying to see if there are any good docs on these values...

legraphista replied:
I'm unsure what `intermediate_size` is either. For `Qwen1.5-MoE-A2.7B` the values look correct, that's why I'm inclined to say that `intermediate_size = 18944` could have been a mistake. 

As for storing `shared_expert_intermediate_size`, I see no specific slot for it, apart from overriding `LLM_KV_FEED_FORWARD_LENGTH` which might not be the best option (even tho' it's unused)

CISC replied:
`Qwen2MoE` docs on HF, not very helpful:
* **intermediate_size** (int, optional, defaults to 5632) — Dimension of the MLP representations.
* **moe_intermediate_size** (int, optional, defaults to 1408) — Intermediate size of the routed expert.
* **shared_expert_intermediate_size** (int, optional, defaults to 5632) — Intermediate size of the shared expert.

CISC replied:
> As for storing `shared_expert_intermediate_size`, I see no specific slot for it, apart from overriding `LLM_KV_FEED_FORWARD_LENGTH` which might not be the best option (even tho' it's unused)

You would have to make a new one.

legraphista replied:
You'd probably want both values captured for the future, in case a model comes along where **ffn_*_exps** and **ffn_*_shexp** shapes are independently driven, but in this case (and I think for a lot of other cases too) they are in sync

CISC replied:
Aha, take a look here:
https://github.com/huggingface/transformers/blob/main/src/transformers/models/qwen2_moe/modeling_qwen2_moe.py#L860-L865

`57B-A14B` is a sparse model and `intermediate_size` is then overridden by `shared_expert_intermediate_size`:
https://github.com/huggingface/transformers/blob/main/src/transformers/models/qwen2_moe/modeling_qwen2_moe.py#L804

legraphista replied:
> > As for storing `shared_expert_intermediate_size`, I see no specific slot for it, apart from overriding `LLM_KV_FEED_FORWARD_LENGTH` which might not be the best option (even tho' it's unused)
> 
> You would have to make a new one.

I'm unsure if this change should be made within the scope of this PR. I could open another PR referencing this one's discussions.

legraphista replied:
> 57B-A14B is a sparse model and intermediate_size is then overridden by shared_expert_intermediate_size

I see, so technically, setting `LLM_KV_FEED_FORWARD_LENGTH` for `57B-A14B` to `shared_expert_intermediate_size` was a correct solution all along

CISC replied:
Yeah, none of them have `mlp_only_layers` yet though.

legraphista replied:
@CISC What do you think the next steps should be?

We can leave this PR as is since it works, and then start a new one with a proper impl?

Otherwise, adding `mlp_only_layers (int[])` and `decoder_sparse_step: int` to the config (and converter) is probably out of scope.

CISC replied:
I think adding a `shared_expert_feed_forward_length` is within the scope of this PR.

The sparse layer stuff should be another PR.

legraphista replied:
Ok, would you like to simply override `LLM_KV_FEED_FORWARD_LENGTH` with `shared_expert_intermediate_size` or create a new one, say `LLM_KV_SHARED_EXPERT_FEED_FORWARD_LENGTH`?

CISC replied:
The latter since you need to keep the original value for the other PR.

compilade replied:
> `mlp_only_layers: int[]`, and `decoder_sparse_step: int`

@legraphista I've had to deal with models which have both MoE layers and MLP-only layers in #7531 (Jamba). A new metadata key-value pair is not needed. The easy way is to [check for the presence of `layers.ffn_gate_inp`](https://github.com/ggerganov/llama.cpp/blob/33425a7e1ed366082a2dbf64f2485531471515e0/llama.cpp#L7756), which is only there on MoE layers, and when building the compute graph, [`build_llama`](https://github.com/ggerganov/llama.cpp/blob/963552903f51043ee947a8deeaaa7ec00bc3f1a4/llama.cpp#L7502) can be a good inspiration for how to check this per-layer.

legraphista commented:
i've updated the `GGUFWriter`, `convert-hf-to-gguf.py` and `llama.cpp` to add support for `LLM_KV_SHARED_EXPERT_FEED_FORWARD_LENGTH`

CISC commented on llama.cpp:
```diff
@@ -360,21 +361,22 @@ static const std::map<llm_kv, const char *> LLM_KV_NAMES = {
+    { LLM_KV_POOLING_TYPE ,                     "%s.pooling_type"                      },
+    { LLM_KV_LOGIT_SCALE,                       "%s.logit_scale"                       },
 
     { LLM_KV_ATTENTION_HEAD_COUNT,          "%s.attention.head_count"             },
```
You might have to align all of them, before and after, up to @ggerganov

CISC reviewed (approved):
LGTM

CISC commented on llama.cpp:
```diff
@@ -5019,6 +5025,11 @@ static void llm_load_print_meta(llama_model_loader & ml, llama_model & model) {
         LLAMA_LOG_INFO("%s: rope_yarn_log_mul    = %.4f\n",   __func__, hparams.rope_yarn_log_mul);
     }
+
+    if (model.arch == LLM_ARCH_QWEN2MOE) {
+        LLAMA_LOG_INFO("%s: n_ff_exp         = %d\n",     __func__, hparams.n_ff_exp);
+        LLAMA_LOG_INFO("%s: n_ff_shared_exp  = %d\n",     __func__, hparams.n_ff_shared_exp);
+    }
```
Ooof, this is broken!

`llm_load_print_meta` is called before `llm_load_tensors` where these params are set, so will always print 0.

It is not a problem introduced in this PR though, this affects others too, like `DeepSeekV2`.

legraphista replied:
That's very strange. When I run `./main` to test & debug, i get the following log snippet:
```
llm_load_print_meta: general.name     = Qwen2-57B-A14B-Instruct
llm_load_print_meta: BOS token        = 151643 '<|endoftext|>'
llm_load_print_meta: EOS token        = 151645 '<|im_end|>'
llm_load_print_meta: PAD token        = 151643 '<|endoftext|>'
llm_load_print_meta: LF token         = 148848 'ÄĬ'
llm_load_print_meta: EOT token        = 151645 '<|im_end|>'
llm_load_print_meta: n_ff_exp         = 2560
llm_load_print_meta: n_ff_shared_exp  = 20480
```

suggesting that the data is loaded before printing

CISC replied:
Never mind me, I'm talking nonsense :)

compilade commented on gguf-py/gguf/constants.py:
```diff
@@ -33,21 +33,22 @@ class General:
+        LEADING_DENSE_BLOCK_COUNT         = "{arch}.leading_dense_block_count"
+        FEED_FORWARD_LENGTH               = "{arch}.feed_forward_length"
+        EXPERT_FEED_FORWARD_LENGTH        = "{arch}.expert_feed_forward_length"
+        SHARED_EXPERT_FEED_FORWARD_LENGTH = "{arch}.shared_expert_feed_forward_length"
```
I wonder, (since this is a new key and can still be renamed) should this be `EXPERT_SHARED_FEED_FORWARD_LENGTH` and `{arch}.expert_shared_feed_forward_length` to match the naming of `EXPERT_SHARED_COUNT`?

This way, anything related to experts would still start with "expert_".

But maybe this is fine as-is too. What do you think?

(Side note: `EXPERT_SHARED_COUNT` seems to be used for a similar purpose in DeepSeekV2. In the case of Qwen2-57B-A14B, I think this could be `8` (`20480/2560 = 8`), but the shared expert feed-forward length seems more general (which is good))

legraphista replied:
> I wonder, (since this is a new key and can still be renamed) should this be `EXPERT_SHARED_FEED_FORWARD_LENGTH` and `{arch}.expert_shared_feed_forward_length` to match the naming of `EXPERT_SHARED_COUNT`?
> 
> This way, anything related to experts would still start with "expert_".
> 
> But maybe this is fine as-is too. What do you think?

I have no preference for the naming of the key. @ggerganov just let me know and I will force-push the update with the changed key name `SHARED_EXPERT_FEED_FORWARD_LENGTH` -> `EXPERT_SHARED_FEED_FORWARD_LENGTH`

ggerganov replied:
Yes `EXPERT_SHARED_FEED_FORWARD_LENGTH` fits better into the naming convention because it reuses an existing prefix

legraphista replied:
fixed in https://github.com/ggerganov/llama.cpp/compare/7fbb7a1b9dcdf57096d0d3e4fc517291ac6f8a23..d9452267a0069a2873f4855cfc0b671b31704889

ggerganov commented on llama.cpp:
```diff
@@ -1840,6 +1842,7 @@ struct llama_hparams {
     uint32_t n_lora_q = 0;
     uint32_t n_lora_kv = 0;
     uint32_t n_ff_exp = 0;
+    uint32_t n_ff_shared_exp = 0;
```
```suggestion
    uint32_t n_ff_shexp = 0;
```

legraphista replied:
force pushed update for the name change.
diff here https://github.com/ggerganov/llama.cpp/compare/559b77819f7f6842a0dd467c5f536aeaed7433bd..46777ddaf370ec6f6b03d9a9d1797269855e4b30

CISC replied:
Should `n_ff_shexp` be checked here as well btw?
https://github.com/ggerganov/llama.cpp/blob/559b77819f7f6842a0dd467c5f536aeaed7433bd/llama.cpp#L1893

CISC commented on llama.cpp:
```diff
@@ -5805,16 +5816,17 @@ static bool llm_load_tensors(
-                        layer.ffn_gate_shexp = ml.create_tensor(ctx_split, tn(LLM_TENSOR_FFN_GATE_SHEXP, "weight", i), {n_embd,   n_ff});
-                        layer.ffn_down_shexp = ml.create_tensor(ctx_split, tn(LLM_TENSOR_FFN_DOWN_SHEXP, "weight", i), {  n_ff, n_embd});
-                        layer.ffn_up_shexp   = ml.create_tensor(ctx_split, tn(LLM_TENSOR_FFN_UP_SHEXP,   "weight", i), {n_embd,   n_ff});
+                        layer.ffn_gate_shexp = ml.create_tensor(ctx_split, tn(LLM_TENSOR_FFN_GATE_SHEXP, "weight", i), {n_embd,   n_ff_shared_exp});
+                        layer.ffn_down_shexp = ml.create_tensor(ctx_split, tn(LLM_TENSOR_FFN_DOWN_SHEXP, "weight", i), {  n_ff_shared_exp, n_embd});
+                        layer.ffn_up_shexp   = ml.create_tensor(ctx_split, tn(LLM_TENSOR_FFN_UP_SHEXP,   "weight", i), {n_embd,   n_ff_shared_exp});
```
nit: rename this variable too?

legraphista replied:
fixed here https://github.com/ggerganov/llama.cpp/compare/46777ddaf370ec6f6b03d9a9d1797269855e4b30..7fbb7a1b9dcdf57096d0d3e4fc517291ac6f8a23#diff-150dc86746a90bad4fc2c3334aeb9b5887b3adad3cc1459446717638605348efL5825-R5826

legraphista committed:
fix: QWEN2MOE support for expert_feed_forward_length

previously, expert ff was taken from n_ff (intermediate size) but it is now properly taken from LLM_KV_EXPERT_FEED_FORWARD_LENGTH

n_ff_exp and n_ff_shexp are now properly calculated

CISC reviewed (approved):
Good to go! :)

ggerganov commented:
We can merge after @compilade approves
